// docpad-plugin-grunt/out/grunt.plugin.js
// Generated by CoffeeScript 1.6.3
(function() {
  var __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports = function(BasePlugin) {
    var GruntPlugin;
    return GruntPlugin = (function(_super) {
      __extends(GruntPlugin, _super);

      GruntPlugin.prototype.name = 'grunt';

      GruntPlugin.prototype.config = {
        writeAfter: [],
        writeBefore: false,
        renderBefore: false,
        renderAfter: false,
        generateBefore: false,
        generateAfter: false,
        populateCollectionsBefore: false,
        populateCollections: false
      };

      function GruntPlugin() {
        GruntPlugin.__super__.constructor.apply(this, arguments);
        this.safeps = require('safeps');
        this.path = require('path');
        this.glob = require('glob');
        this;
      }

      GruntPlugin.prototype.writeBefore = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().writeBefore || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.writeAfter = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().writeAfter || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.renderBefore = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().renderBefore || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.renderAfter = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().renderAfter || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.generateBefore = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().generateBefore || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.generateAfter = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().generateAfter || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.populateCollectionsBefore = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().populateCollectionsBefore || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.populateCollections = function(opts, next) {
        var tasks;
        if (tasks = this.getConfig().populateCollections || false) {
          this.processGrunt(tasks, opts, next);
        } else {
          return next();
        }
        return this;
      };

      GruntPlugin.prototype.processGrunt = function(tasks, opts, next) {
        var command, err, files, gruntPath, rootPath, task, _i, _len, _ref;
        rootPath = this.docpad.getConfig().rootPath;
        files = this.glob.sync('**/grunt-cli/bin/grunt', {
          cwd: rootPath,
          nosort: true
        });
        if (gruntPath = files[0] || false) {
          command = [this.path.join(rootPath, gruntPath)];
          _ref = tasks || [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            task = _ref[_i];
            command.push(task);
          }
          this.safeps.spawn(command, {
            cwd: rootPath,
            output: true
          }, next);
        } else {
          err = new Error('Could not find the Grunt command line interface.');
          return next(err);
          err;
        }
        return this;
      };

      return GruntPlugin;

    })(BasePlugin);
  };

}).call(this);
